libvast/builtins/operators/shell.cpp

```diff
@@ -0,0 +1,292 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/as_bytes.hpp>
+#include <vast/chunk.hpp>
+#include <vast/concept/parseable/string/quoted_string.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/error.hpp>
+#include <vast/logger.hpp>
+#include <vast/pipeline.hpp>
+#include <vast/plugin.hpp>
+#include <vast/si_literals.hpp>
+
+#include <boost/asio.hpp>
+#include <boost/process.hpp>
+#include <caf/detail/scope_guard.hpp>
+
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace bp = boost::process;
+
+namespace vast::plugins::shell {
+namespace {
+
+using namespace vast::binary_byte_literals;
+
+/// The block size when reading from the child's stdin.
+constexpr auto block_size = 16_KiB;
+
+/// Wraps the logic for interacting with a child's stdin and stdout.
+class child {
+public:
+  static auto make(std::string command) -> caf::expected<child> {
+    auto result = child{std::move(command)};
+    try {
+      auto exit_handler = [](int exit, std::error_code ec) {
+        VAST_DEBUG("child exited with code {}: {}", exit, ec.message());
+      };
+      result.child_ = bp::child{
+        result.command_,
+        bp::std_out > result.stdout_,
+        bp::std_in < result.stdin_,
+        bp::on_exit(exit_handler),
+      };
+    } catch (const bp::process_error& e) {
+      return caf::make_error(ec::filesystem_error, e.what());
+    }
+    return result;
+  }
+
+  auto reading() -> bool {
+    return child_.running() && not stdout_.eof();
+  }
+
+  auto writing() -> bool {
+    return child_.running() && not stdin_.eof();
+  }
```

Comment:
```suggestion
  auto reading() const -> bool {
    return child_.running() && not stdout_.eof();
  }

  auto writing() const -> bool {
    return child_.running() && not stdin_.eof();
  }
```

Reply:
Not possible, `child_.running()` is non-const. It's a Boost Process thing, don't ask me why.